futures-async-macro/src/lib.rs

```diff
@@ -0,0 +1,408 @@
+                                ::futures::core_reexport::option::Option::Some(e) => e,
+                                ::futures::core_reexport::option::Option::None => break,
+                            }
+                        };
+
+                        #body
+                    }
+                }}
+            }
+            Stream => {
+                // Basically just expand to a `poll` loop
+                syn::parse_quote! {{
+                    let mut __pinned = #expr;
+                    #label
+                    loop {
+                        let #pat = {
+                            match ::futures::async_stream::poll_next_with_tls_context(unsafe {
+                                    ::futures::core_reexport::pin::Pin::new_unchecked(&mut __pinned)
+                                })
+                            {
+                                ::futures::core_reexport::task::Poll::Ready(e) => {
+                                    match e {
+                                        ::futures::core_reexport::option::Option::Some(e) => e,
+                                        ::futures::core_reexport::option::Option::None => break,
+                                    }
+                                }
+                                ::futures::core_reexport::task::Poll::Pending => {
+                                    yield ::futures::core_reexport::task::Poll::Pending;
+                                    continue
+                                }
+                            }
+                        };
+
+                        #body
+                    }
+                }}
+            }
+            Closure => return outside_of_async_error!(expr, "#[for_await]"),
+        }
+    }
+
+    /// Expands `yield expr` in `async_stream` scope.
+    fn expand_yield(&self, expr: ExprYield) -> ExprYield {
+        if self.0 != Stream { return expr }
+
+        let ExprYield { attrs, yield_token, expr } = expr;
+        let expr = expr.map_or_else(|| quote!(()), ToTokens::into_token_stream);
+        let expr = syn::parse_quote! {
+            ::futures::core_reexport::task::Poll::Ready(#expr)
+        };
+        ExprYield { attrs, yield_token, expr: Some(Box::new(expr)) }
+    }
+
+    /// Expands `async_stream_block!` macro.
+    fn expand_macro(&mut self, mut expr: ExprMacro) -> Expr {
+        if expr.mac.path.is_ident("async_stream_block") {
+            let mut e: ExprCall = syn::parse(async_stream_block(expr.mac.tts.into())).unwrap();
+            e.attrs.append(&mut expr.attrs);
+            Expr::Call(e)
+        } else {
+            Expr::Macro(expr)
+        }
+    }
+
+    /// Expands `expr.await` in `async_stream` scope.
+    ///
+    /// It needs to adjust the type yielded by the macro because generators used internally by
+    /// async fn yield `()` type, but generators used internally by `async_stream` yield
+    /// `Poll<U>` type.
+    fn expand_await(&mut self, expr: ExprField) -> Expr {
```

Comment:
It is not possible to replace `.await` in macros such as `assert!(foo.await)` with this.

Full support for this can not be done without compiler support, but it can support for some kinds of macros (however, this is not necessarily preferable, as a macro may call an async block etc): https://github.com/taiki-e/futures-rs/commit/fb33504229db314fb97759c3170ee3ad73785c08

Related: https://github.com/Nemo157/embrio-rs/issues/14